### ISSUE TITLE
improving keyboard layout  managing

### ISSUE THREAD

**Issue by michaelDCurran** [nvaccess/test_nvdaTracImport#143]
**Reported by aleksey s on 9 Apr 2008 18:24 UTC**
when nvda loads keyboard layout, it must check and log how many scripts were not mapped, and probably log their names too,  to be easier for layout creators to update their files. 
also some combobox needed for user to can change keyboard layout with out editing nvda.ini file. nvda might search *.kbd files and put to combobox all diferent parts after underline. so keyboard layouts might be named e.g. laptop, mick's laptop etc.

**Comment by michaelDCurran** [comment]
**Comment by mdcurran on 30 Apr 2008 05:20 UTC**
Added support for selecting keyboard layout from the keyboard settings dialog in svn r1989. See trac ticket #60 for more info. In regards to logging bad key names, this shouldn't be too hard. Though I think exceptions are already log for any script name that doesn't exist, though I guess this may b a bit unclear as to what that really means.

**Comment by michaelDCurran** [comment]
**Comment by mdcurran on 17 Nov 2008 22:44 UTC**
Seems that NVDA does log any key binding that failed because the script did not exist. So I think we're done with this ticket.

**Comment by michaelDCurran** [comment]
**Comment by mdcurran on 17 Nov 2008 22:59 UTC**
NVDA is much different to other screen readers in its dictionary implementation. Ordering of entries is very important. If we ordered by comment or anything else other than the actual file order, then the user may be adding entries that may not work because they could be overridden by ones that appear above their entry.

**Comment by michaelDCurran** [comment]
**Comment by mdcurran on 17 Nov 2008 23:00 UTC**
Please ignore the last comment, my mystake.

**Comment by michaelDCurran** [comment]
**Comment by aleksey_s on 22 Nov 2008 22:00 UTC**
i requested logging script names which exists but not mapped to key as well as warning when one keystroke is used in different scripts in same appmodule. this isn't yet implemented.

**Comment by michaelDCurran** [comment]
**Modified by jteh on 18 May 2009 05:27 UTC**

**Comment by michaelDCurran** [comment]
**Comment by MHameed on 25 Feb 2013 19:49 UTC**
gestures manager:
With more addons being created and being circulated, the likelyhood of clashing shortcut keys is increasing.
Would it be possible for an appmodule/globalModule to register its functions (without any assigned keys), then the user could browse the functionality and the unassigned/assigned/reassign keys.
This would be simular to Orca keybindings, i.e. a treeview for each group, when the group is expanded one could browse the available functionality, and reset to default, reassign to a different key etc.
This would also aid functionality/shortcut discoverability.

**Comment by michaelDCurran** [comment]
**Comment by jteh on 26 Feb 2013 06:05 UTC**
Replying to [MHameed](comment:7):
> Would it be possible for an appmodule/globalModule to register its functions (without any assigned keys)
Imo, this isn't a good idea. I recognise the potential for clashes, but not having any bindings at all is very unfriendly for users. There could be a huge number of commands that need to be bound.

There is definitely a need for a GUI to manage custom gesture binding. This is covered by #1532, though implementing it is very complicated due to the hugely dynamic way NVDA handles bindings.

**Comment by michaelDCurran** [comment]
**Comment by nvdakor on 25 Jul 2014 12:53 UTC**
Hi,
Now that #1532 has been implemented, I think this ticket could be closed unless there are issues to be solved.
Thanks.

**Comment by michaelDCurran** [comment]
**Comment by jteh on 28 Jul 2014 06:45 UTC**
This isn't really relevant any more, since we don't use kbd files now and we also have #1532.
